Automate rotation of the service CA

marun opened the pull request:
This PR is a fork of #42.

Co-authored-by: Matt Rogers <mrogers@redhat.com>
Co-authored-by: Stanislav Laznicka <slaznick@redhat.com>

marun commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,71 @@
+
+// RotateSigningCA creates a new CA with a set of cross-signed interim CAs to allow for graceful rollover.
+// Returns new CA pem, new CA key pem and the full client CA bundle.
+func RotateSigningCA(currentCert *x509.Certificate, currentKey *rsa.PrivateKey) ([]byte, []byte, []byte, error) {
```
@stlaz Does this look ok?

marun replied:
Done

stlaz commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,71 @@
+	signedByCurrentCert := &x509.Certificate{Raw: rawSignedByCurrentCert}
+
+	// Create the full CA bundle comprising all 4 certs
+	// TODO(marun) Is the ordering of certs in the bundle important?
```
It shouldn't.

stlaz commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,71 @@
+	// TODO(marun) Is the ordering of certs in the bundle important?
+	bundleCerts := []*x509.Certificate{
+		currentCert,
+		signedByNewCert,
```
This cert may not even be needed since we have the old CA in the trust bundle. It would be nice to have an e2e test which proves this.

marun replied:
I think the [design doc](https://docs.google.com/document/d/1VqrIERs30M9EyPaqcO43wcSXHXvfe_Ao2W0ThDIohlM/edit) was presuming a signing scheme that relied on intermediates, but since that's not the strategy used by this operator (all service certs are signed by the current 'root'), the only intermediate that matters is #4 (including it with service certs supports support clients that haven't yet updated to the newest ca bundle). That implies that neither intermediate cert needs to be part of the bundle since the old and new certs (`#1` and `#2`) are by themselves sufficient to validate `X` and `Y`. `#3` is unnecessary.

stlaz commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	// TODO(marun) Is it necessary to append signedByCurrentCert to
+	// newCertConfig.Certs? PR #42 was doing so.
```
See the above comment, I drawed this on our whiteboard and if I am correct, we wouldn't need the `signedByNewCert` instead

marun replied:
I've come to the conclusion that this was never necessary (not sure why #42 was doing so) - CA cert `#4` needs to be included with each new service cert rather than being included with the signing CA.

stlaz replied:
Re-reading the doc and thinking about it once again, that does sound like a correct assumption given the #4 is distributed by the server and #3 will never really sign anything.

marun replied:
Going back to this comment, how do you imagine that `#4` is distributed by the server if it is not appended to generated serving certs?

marun commented on pkg/operator/sync_common.go:
```diff
@@ -78,31 +83,97 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
+}
+
+// crossSign generates a new CA cert and key
+func crossSign(caCert *x509.Certificate, keyData []byte) ([]byte, []byte, []byte, error) {
```
I've inlined this function for now. I'm less concerned with getting the code as neat as possible as getting it functional, and the next tricky thing to tackle appears to be plumbing through the use of the intermediate cert (needs to be appended to service certs if present).

stlaz replied:
That's fine

marun commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,71 @@
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	signedByNewCert := &x509.Certificate{Raw: rawSignedByNewCert}
```
Updated to use ParseCertificate as per your request.  Also, this method now delegates conversion to PEM to the `SigningCA` type to simplify testing.

marun commented on pkg/operator/util/ca_test.go:
```diff
@@ -0,0 +1,186 @@
+
+// TestRotateSigningCA validates that service certs signed by pre- and
+// post-rotated CAs can be validated by both the new and old bundles.
+func TestRotateSigningCA(t *testing.T) {
```
@stlaz This test has been rewritten to validate the revised rotation strategy.

marun commented:
TODO
- [ ] Ensure an intermediate cert (`#4`) is appended to generated signing certs if present
- [ ] Rewrite e2e to validate client trust as per the updated unit test.

enj commented:
/uncc

@stlaz @sttts @mfojtik

stlaz commented on pkg/operator/util/ca_test.go:
```diff
@@ -0,0 +1,186 @@
+		if err := srv.ServeTLS(ln, "", ""); err != nil && err != http.ErrServerClosed {
+			t.Fatalf("ServeTLS failed: %v", err)
+		}
+	}()
```
Go merely passes the `srv.Addr` to `net.Listen` so the outcome should be the same - https://golang.org/src/net/http/server.go?s=97796:97864#L3098

I agree that this is a good candidate for an e2e test, but the unit test should eventually have just the verification part (which is what happens on that https server anyway, but I get your point) :+1:

marun replied:
The test starts a server on a port guaranteed not to be already bound to ensure reliable execution in a test environment, which is achieved by using an address of `127.0.0.1:0`. As you say, it should be possible to set `srv.Addr` to that address and have the server start on a random port. However, `srv.Addr` is not updated once `ListenAndServeTLS` starts the server - the bound port is only set on the listener internal to that call chain. Hence the reason I'm starting the listener in advance - to enable reliable binding to a random port _and_ retrieval of that port to construct the address the client should target.

In my previous comment I said `I think it would be fine to switch to simple verification before merge`, and by that I meant replacing the https-based testing with cert verification. 

If we wanted to retain https-based testing, I'm not sure why it would be desirable to do that in e2e since verifying the cert chain of trust doesn't depend on a cluster. I'm a firm believer in running tests in the lowest-cost environment possible, and unit testing is about as cheap as we get (this test as currently written has sub-second execution time).

marun replied:
I've updated the unit test to perform cert validation, and rewritten the e2e test to validate the trust with http server+client. The e2e test is not yet passing, I'm working on debugging the operator locally.

marun commented:
This PR is dependent on https://github.com/openshift/library-go/pull/560 and https://github.com/openshift/api/pull/482.

The code is functionally complete, but the architecture of this operator make it very challenging to debug locally. For now I'm going to rely on CI.

stlaz commented on pkg/operator/sync_common.go:
```diff
@@ -78,31 +85,88 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
+			return false, err
+		}
+		d := secret.Data
+		d[corev1.TLSCertKey], d[corev1.TLSPrivateKeyKey], d[api.IntermediateDataKey], d[api.BundleDataKey], err = signingCA.GetPEMBytes()
```
You swapped Intermediate and Bundle keys here

marun replied:
Good catch!  Done.

stlaz commented on pkg/operator/util/ca.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// GetPEMBytes returns PEM-encodings of the CA cert, key, bundle and intermediate CA cert.
+func (ca *SigningCA) GetPEMBytes() ([]byte, []byte, []byte, []byte, error) {
```
Inline this within `RotateSigningCA()`, the struct above is single-purpose, you're only carrying the bits over.

marun replied:
Externalizing conversion to this method simplifies unit testing since conversion to PEM is unnecessary for the purposes of cert verification. Or would you like to see the exported method `RotateSigningCA` return PEM and an internal method `rotateSigningCA` that returns `SigningCA`?

stlaz commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -313,11 +322,17 @@ func getServingCert(dnsSuffix string, ca *crypto.CA, service *corev1.Service) (*
-func toRequiredSecret(dnsSuffix string, ca *crypto.CA, service *corev1.Service, secretCopy *corev1.Secret) error {
-	servingCert, err := getServingCert(dnsSuffix, ca, service)
+func toRequiredSecret(dnsSuffix string, ca *crypto.CA, intermediateCACert *x509.Certificate, service *corev1.Service, secretCopy *corev1.Secret) error {
+	servingCert, err := getServingCert(dnsSuffix, ca, intermediateCACert, service)
```
The original PR was not adding the intermediate to the serving cert. I think commonly the server just sends its whole chain. I wonder whether this step would actually make a server which would otherwise send just its own cert send both or not.

marun replied:
The server only sends the certs that are supplied to it. Typically that won't include an entire trust chain, but only the certs required for the client to verify from what the client already knows (i.e. a ca bundle). The goal here is ensuring that a server using a serving cert will send clients `#4` and `Y` so that unrefreshed clients will be able to verify `Y` against an old bundle that contains `#1` but is missing `#2` (because `Y` -> `#4` -> `#1`).

I talked to Matt yesterday and confirmed that the original PR was in error in not adding the intermediate (`#4`) to new serving certs. The design doc even explicitly says it is required - `Distribute #4 with the new service-certs (appended to PEM)`.

I've also confirmed this in the unit testing - try removing [the addition of the intermediate cert](https://github.com/openshift/service-ca-operator/pull/73/files#diff-c482e5ba51661f6d2b95bcc56e0cd468R59) and the `Server rotated` test scenario will fail as expected.

stlaz commented on pkg/controller/servingcert/starter/starter.go:
```diff
@@ -31,6 +35,18 @@ func StartServiceServingCertSigner(ctx *controllercmd.ControllerContext) error {
 	}
 
+	var intermediateCert *x509.Certificate
+	if len(config.IntermediateCertFile) > 0 {
```
Nvm, this is wrong

marun commented on glide.diff:
```diff
@@ -0,0 +1,72 @@
+diff --no-dereference -N -r current/vendor/github.com/openshift/api/servicecertsigner/v1alpha1/types.go updated/vendor/github.com/openshift/api/servicecertsigner/v1alpha1/types.go
```
This was meant to be pushed - the commits associated with it are marked `DO NOT MERGE` because as per [my previous comment](https://github.com/openshift/service-ca-operator/pull/73#issuecomment-543607442) changes need to merge to deps before this PR will be mergeable.

marun commented:
@stlaz The intermediate cert (`#4`) is only required to be included in serving certs - this ensures that clients without the newest ca bundle can verify trust. Clients with the updated ca bundle containing `#1` and `#2` don't need `#4` since both `X` and `Y` can be validated without it.

marun commented:
@stlaz Thank you for your diligent review, much appreciate!

> I'm not sure about the intermediate CA split from the bundle.

The intermediate cert (`#4`) is only required to be included in serving certs - this ensures that clients without the newest ca bundle can verify trust. Clients with the updated ca bundle containing `#1` and `#2` don't need `#4` since both `X` and `Y` can be validated without it.

marun commented:
/retest

marun commented:
/retest

marun commented:
/hold

marun commented:
Now that [the enhancement](https://github.com/openshift/enhancements/pull/83) as merged, this PR is ready for review. I've merged #77 (forced rotation) into this PR since it ended up cleaning up the code in ways that I thought were more reviewable.

As previously commented, openshift/library-go#560 and openshift/api#482 need to merge before this PR will be mergeable. This PR is vendoring those PR branches instead of master until master includes the necessary changes. 

Once the dep PRs merge, should a bump be performed separately from this PR?

marun commented:
/retest

stlaz commented on test/e2e/e2e_test.go:
```diff
@@ -11,15 +11,23 @@ import (
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
```
nit: `kruntime`?

marun replied:
Done

`pkgruntime` is the alias I've commonly seen in k/k and related upstream projects, I guess `kruntime` is the openshift version?

stlaz commented on pkg/operator/sync_common.go:
```diff
@@ -78,31 +82,79 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
+	_, mod, err := resourceapply.ApplySecret(client, eventRecorder, secret)
+
+	if err == nil && len(rotationMsg) > 0 {
+		eventRecorder.Eventf("RotatedServiceCA", rotationMsg)
```
nit: `ServiceCARotated` matches better other event names

marun replied:
Done

stlaz commented on test/util/rotate.go:
```diff
@@ -0,0 +1,171 @@
+	}()
+	defer func() {
+		err = srv.Close()
+		if err != nil {
```
nit: `if err := srv.Close(); err != nil`, unless you meant to assign to the outer scope error

marun replied:
Done

stlaz commented on pkg/operator/rotate_test.go:
```diff
@@ -0,0 +1,280 @@
+
+	// Create a secret that has been force-rotated for a reason
+	forceRotatedSigningSecret := newSigningSecret.DeepCopy()
+	recordForcedRotationReason(forceRotatedSigningSecret, "42")
```
lol

marun commented on pkg/operator/sync_common.go:
```diff
@@ -177,7 +177,9 @@ func manageConfigMapCABundleControllerDeployment(client appsclientv1.AppsV1Inter
 	required.Spec.Template.Spec.Containers[0].Image = os.Getenv("CONTROLLER_IMAGE")
-	required.Spec.Template.Spec.Containers[0].Args = append(required.Spec.Template.Spec.Containers[0].Args, fmt.Sprintf("-v=%s", options.Spec.LogLevel))
+	if len(options.Spec.LogLevel) > 0 {
+		required.Spec.Template.Spec.Containers[0].Args = append(required.Spec.Template.Spec.Containers[0].Args, fmt.Sprintf("-v=%s", options.Spec.LogLevel))
```
Done

marun commented on manifests/03_operator.cr.yaml:
```diff
@@ -6,3 +6,5 @@ metadata:
 spec:
   managementState: Managed
+  # Configure the log level of managed deployments
+  logLevel: "4"
```
> With the change from previous comment, you should not need this, I hope. However, it changes the default loglevel to 2.

Done
 
> Isn't it weird we're running with debug loglevel by default?

I guess so. I wasn't intending to affect the logging behavior in this PR, not sure why previous maintainers decided to set 4 as the default.

marun replied:
I've updated the verbosity of logging of rotation initiation to `2`, since it is infrequent and is something that the user should be informed of even without debug logging enabled.

marun commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -227,7 +231,21 @@ func (sc *serviceServingCertController) issuedByCurrentCA(secret *corev1.Secret)
+	// generated before serving cert generation was updated to include the field.
+	checkKeyIdentifierChaining := len(caSubjectKeyId) > 0
+	if checkKeyIdentifierChaining {
+		return bytes.Compare(certAuthorityKeyId, caSubjectKeyId) == 0
```
Done

marun commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -301,23 +319,34 @@ func toBaseSecret(service *corev1.Service) *corev1.Secret {
 
-func getServingCert(dnsSuffix string, ca *crypto.CA, service *corev1.Service) (*crypto.TLSCertificateConfig, error) {
-	dnsName := service.Name + "." + service.Namespace + ".svc"
+func GetServingCert(dnsSuffix string, ca *crypto.CA, intermediateCACert *x509.Certificate, objMeta *metav1.ObjectMeta) (*crypto.TLSCertificateConfig, error) {
```
Done

marun commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -301,23 +319,34 @@ func toBaseSecret(service *corev1.Service) *corev1.Secret {
 	if err != nil {
 		return nil, err
 	}
+	if len(servingCert.Certs) > 1 {
```
Done.

It's not really a precaution, it's just discontinuing the practice of unnecessarily including the issuing cert. Maybe this is desirable if the issuing cert were an intermediate that was not shipped with the CA bundle, but that isn't the case for service CA. I'm guessing this is a consequence of library-go serving many different use cases, one of which requires this behavior.

sttts replied:
The primary CA is not trusted (i.e. signed by some root CA which everybody trusts), right? So every client should have access to the CA bundle already, and hence this change should not have any consequence?

marun commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// getPEMBytes returns PEM-encodings of the CA cert, key, bundle and intermediate CA cert.
+func (ca *signingCA) getPEMBytes() ([]byte, []byte, []byte, []byte, error) {
```
I've refactored this method to `updateSigningSecret` which avoids the need to return so many values.

marun commented on pkg/operator/sync_common.go:
```diff
@@ -78,31 +82,79 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
+			return false, err
+		}
+	} else {
+		rotationMsg, err = rotateSigningSecret(existing, existingCert, rawUnsupportedServiceCAConfig)
```
Why is it necessary to write the changes to `secret`? In the case of an existing secret, the expected form was previously only used if the secret did not exist or had a missing/malformed value for `tls.crt`.

`existing` is the result of a direct client call rather than being sourced from an informer, so there is no cache involved.

marun commented:
@stlaz Thank you for your diligent review of this admittedly huge PR!

I've addressed most of your comments (with the exception of [one](https://github.com/openshift/service-ca-operator/pull/73#discussion_r341094150) that I would appreciate further feedback on) in a fixup commit to simplify reviewing. The exception was for some of the minor supporting commits, which I amended.

Note that the `Fixup 2` commit addresses an oversight that seems obvious in hindsight. Rotating a CA near expiry (e.g. for a 4.1 cluster upgraded to 4.3 very close to a year after deployment) previously might have resulted in the old CA not being trusted long enough for clients and servers to refresh. `Fixup 2` ensures that the old CA remains trusted for a minimum duration.

@sttts ^ I think you touched on this in your review of the enhancement, let me know what you think about the proposed fix.

marun commented:
/retest

sttts commented on pkg/operator/sync_common.go:
```diff
@@ -177,7 +178,7 @@ func manageConfigMapCABundleControllerDeployment(client appsclientv1.AppsV1Inter
 	required := resourceread.ReadDeploymentV1OrDie(v4_00_assets.MustAsset(resourcePath + "deployment.yaml"))
 	required.Spec.Template.Spec.Containers[0].Image = os.Getenv("CONTROLLER_IMAGE")
-	required.Spec.Template.Spec.Containers[0].Args = append(required.Spec.Template.Spec.Containers[0].Args, fmt.Sprintf("-v=%s", options.Spec.LogLevel))
+	required.Spec.Template.Spec.Containers[0].Args = append(required.Spec.Template.Spec.Containers[0].Args, fmt.Sprintf("-v=%d", loglevel.LogLevelToKlog(options.Spec.LogLevel)))
```
This is not only about breakage with empty value, but about the API semantics. Before we wanted numbers as string type, now Normal, Debug, Trace, TraceAll.

Was this just broken before?

sttts replied:
Looking at the API, it was broken.

marun replied:
Is this comment actionable?

sttts replied:
No. You are fixing the feature.

sttts commented on manifests/06_config.yaml:
```diff
@@ -1,10 +0,0 @@
-  managementState: Managed
-  version: 4.0.0
-  logLevel: "4"
-  replicas: 1
```
nit: confusing commit message. This one is a duplicate. The other one existing has `release.openshift.io/create-only: "true"` set.

marun replied:
Done

sttts commented on glide.yaml:
```diff
@@ -15,11 +15,15 @@ import:
-  version: master
+  # TODO(marun) Update to master once this branch has merged
+  version: add-akid-skid-to-certs
+  repo: git@github.com:marun/library-go.git
```
tagged. Should merge any moment.

sttts commented on glide.yaml:
```diff
@@ -15,11 +15,15 @@ import:
-  version: master
+  # TODO(marun) Update to master once this branch has merged
+  version: add-intermediate-cert-file
+  repo: git@github.com:marun/api.git
```
tagged. Should merge any moment.

sttts commented on pkg/controller/api/api.go:
```diff
@@ -7,9 +7,12 @@ import (
+	InjectionDataKey                   = "service-ca.crt"
+	BundleDataKey                      = "ca-bundle.crt"
+	IntermediateDataKey                = "intermediate-ca.crt"
 )
```
oneliner godocs would be helpful

marun replied:
Done

sttts commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -227,7 +231,20 @@ func (sc *serviceServingCertController) issuedByCurrentCA(secret *corev1.Secret)
 	}
 
-	return certs[0].Issuer.CommonName == sc.commonName()
+	if certs[0].Issuer.CommonName != sc.commonName() {
```
this is for backwards compatibility? godoc please.

marun replied:
I'm assuming you meant add a comment, since godoc applies to declarations? Or did you want a change in godoc of the function?

sttts replied:
yes

marun replied:
Done

sttts commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -301,23 +318,34 @@ func toBaseSecret(service *corev1.Service) *corev1.Secret {
 
-func getServingCert(dnsSuffix string, ca *crypto.CA, service *corev1.Service) (*crypto.TLSCertificateConfig, error) {
-	dnsName := service.Name + "." + service.Namespace + ".svc"
+func GetServingCert(dnsSuffix string, ca *crypto.CA, intermediateCACert *x509.Certificate, serviceObjectMeta *metav1.ObjectMeta) (*crypto.TLSCertificateConfig, error) {
```
this is not a getter, but a constructor. NewServingCert

sttts replied:
or MakeServingCert

marun replied:
Done

sttts commented on pkg/controller/servingcert/starter/starter.go:
```diff
@@ -31,6 +37,11 @@ func StartServiceServingCertSigner(ctx *controllercmd.ControllerContext) error {
 		return err
 	}
 
+	intermediateCACert, err := readIntermediateCACert(config.IntermediateCertFile)
```
nit: would prefer to have the `len == 0` logic here. Noop funcs with empty parameters are always suspicious.

marun replied:
I'm afraid I'm not sure I understand what you mean by `Noop funcs with empty parameters`.  I've added a comment indicating that the intermediate cert will only be expected after rotation, does that help?

sttts replied:
would check for the filename to be non-empty here, not inside of a func (which has no clue about the context it is called in).

marun replied:
Done

sttts commented on pkg/controller/servingcert/starter/starter.go:
```diff
@@ -64,3 +77,29 @@ func StartServiceServingCertSigner(ctx *controllercmd.ControllerContext) error {
+	certsPEMBlock, err := ioutil.ReadFile(filename)
+	if os.IsNotExist(err) {
+		klog.V(4).Infof("%q does not exist which indicates that an intermediate certificate was not specified", filename)
+		return nil, nil
```
shouldn't we fall over instead? When can this happen?

marun replied:
An intermediate cert will only exist if rotation has performed. For a new cluster, that will not be true. I've updated the godoc for the function to reflect this, PTAL.

sttts commented on pkg/controller/servingcert/starter/starter.go:
```diff
@@ -64,3 +77,29 @@ func StartServiceServingCertSigner(ctx *controllercmd.ControllerContext) error {
+	if err != nil {
+		return nil, fmt.Errorf("error parsing intermediate cert from %s: %s", filename, err)
+	}
+	if len(certs) > 1 {
```
`!= 1`

marun replied:
Done

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+	}
+	if secret.Annotations == nil {
+		return true
+	}
```
nit: you can skip this if

marun replied:
Will `Annotations` will never be nil on a resource returned from the API?  In unit testing that won't be true unless the map is explicitly initialized, but if that's testing-only I should update initialization of secrets used for testing.

sttts replied:
`nil["foo"]` gives empty string. Doesn't that match your logic here?

marun replied:
I was confused - I added this check some time ago in response to an error I saw but it was clearly unrelated.

Done.

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+// rotateSigningSecret rotates the CA provided by the given secret and updates the secret
+// with the new CA's data.
```
... or does nothing if neither a force rotation takes place nor the old ca expired.

marun replied:
Done

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+
+// rotateSigningSecret rotates the CA provided by the given secret and updates the secret
+// with the new CA's data.
+func rotateSigningSecret(secret *corev1.Secret, currentCACert *x509.Certificate, rawUnsupportedServiceCAConfig []byte) (string, error) {
```
maybeRotateSigningSecret

marun replied:
Done

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+	}
+	key, err := keyutil.ParsePrivateKeyPEM(keyData)
+	if err != nil {
+		return "", err
```
better error context?

marun replied:
Done

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+		return "", err
+	}
+
+	signingCA, err := rotateSigningCA(currentCACert, key.(*rsa.PrivateKey))
```
could this be a non RSA key?

sttts replied:
At those which come from secrets externally should have a type check.

marun replied:
Done

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+// certHalfwayExpired indicates whether half of the cert validity period has elapsed.
+func certHalfwayExpired(cert *x509.Certificate) bool {
+	halfValidPeriod := cert.NotAfter.Sub(cert.NotBefore).Nanoseconds() / 2
+	halfExpiration := cert.NotBefore.Add(time.Duration(halfValidPeriod) * time.Nanosecond)
```
is that `* time.Nanosecond`  really needed? Doesn't Add take a time.Time?

marun replied:
I've removed this function in preference to triggering time-based rotation when the current CA's expiry provides less than the minimum trust duration (~6 months) rather than checking that the CA's is more than half-way through its validity period. This is in keeping with how long the previous CA will be trusted. WDYT?

sttts replied:
it's the same at the end, isn't it?

marun replied:
Effectively the same, but I think it's more clear to using the same constant (`minimumTrustDuration`) for both detection that time-based rotation is required and to ensure that the previous CA remains trusted for at least that same duration.

sttts commented on pkg/operator/rotate.go:
```diff
@@ -0,0 +1,242 @@
+func forceRotationReason(rawUnsupportedServiceCAConfig []byte) (string, error) {
+	serviceCAConfig := &unsupportedServiceCAConfig{}
+	if raw := rawUnsupportedServiceCAConfig; len(raw) > 0 {
+		jsonRaw, err := kyaml.ToJSON(raw)
```
I believe this is JSON already.

marun replied:
Done

sttts commented on pkg/operator/sync_common.go:
```diff
@@ -78,31 +83,79 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
+	rotationMsg := ""
+	if existingCert == nil {
+		// Secret does not exist or lacks the expected cert.
+		err = initializeSigningSecret(secret)
```
`if err := ... ; err !=nil {`

marun replied:
Done

marun commented:
@sttts: Addressed your comments with https://github.com/openshift/service-ca-operator/pull/73/commits/905f5b0251451784567a79b9c4a73b04d53a89c9 and https://github.com/openshift/service-ca-operator/pull/73/commits/fcadd10bc374f5d9e49965347460599327e4c816.

sttts commented:
/lgtm
/approve

marun commented:
/hold cancel
